### PR TITLE
net: Introduction of global protocol handler registry -> netreg (WIP)

### PR DESCRIPTION
**Pull request by haukepetersen**
This PR is opened for discussion and to demonstrate the concept of a global protocol handler/module registry that takes care or keeping lists of PID's.

This code is not compiling (I havn't even tried) and please don't spend much thought on the doxygen... I just want to get feedback on the overall concept.

The main things shown are:
- API and implementation of a global module registry, called `netreg`
- Introduction of a global list of used protocols, called `netmod`
- Simplified `netapi` by throwing out everything related to (un-)registering
- Adjusted `pkt_t` to make use of `netmod`
- Simplified `nomac` by making use of `netreg` (and integrating some thoughts about netdev -> compare to #2297)

Let me know what you think or where I have more explaining to do!

### CODE REVIEW

**Comment by authmillenon** [comment]
In general I like the idea, but I'm conflicted if

- overhead might be increased by this
- making the connection protocol <=> module is such a good idea (think dual-stack with cc3000 e. g.)

**Comment by haukepetersen** [comment]
> - overhead might be increased by this
I think the contrary is true, as the network modules do not have to care about storing PID's anymore -> see removed code in `netapi` and `nomac`...

> making the connection protocol <=> module is such a good idea (think dual-stack with cc3000 e. g.)
You have to think of protocol more as network module, so we could add a module `CC3000_UDP` or something.

**Comment by authmillenon** [comment]
> You have to think of protocol more as network module, so we could add a module CC3000_UDP or something

This will definitely increase code overhead. Instead of: "give me all UDP threads" it would be "give me UDP1, UDP2, UDP3" and ideally (not) with an humongous tree of `#ifdefs` which make the code just not fun anymore. 

 In general a I see a lot of reintroduction and requirements of `#ifdef`s in your rework of the model (not only here). I was under the impression (and already stated this multiple times) however, that one of the major goals of the refactoring was to reduce them, since one of the lessons learned in the transceiver API was that they just don't scale and make the code a pain to read...

**Comment by authmillenon** [comment]
On a more constructive note: how about making the registry a list of (kernel_pid_t, kernel_pid_t, pkt_proto_t) tuples, where the first ist always the lower and the last always the upper layer thread. This way we identify implementations by their PIDs and specifications by an arbitrary number that is independent of the implementation as before.

**Comment by authmillenon** [comment]
There is also the matter of demultipexing identifiers (next header identifiers in IPv6, Ports in UDP/TCP) that is missing from here.

**Comment by haukepetersen** [comment]
Hm, I see it completely on the contrary: For one I don't see any `#ifdefs` popping up, except on one single, central location, namely in `netmod.h`.

I also don't see the overhead: Why would you do something like: give me all UDP threads?

**Comment by authmillenon** [comment]
Ok, UDP might have been a bad example, but take for example Broadcast in IP: "give me all MAC threads".

**Comment by authmillenon** [comment]
It's also still unclear to me: If I have a e.g. a cc110x device with a `nomac` control thread, would I choose `NETMOD_CC110x` or `NETMOD_NOMAC`. If the first is the case first, how would I than differentiate between two cc110x that would run in a setup, where they use to different MAC implementations (since we might want to run experiments like this, this might be a realistic scenario)? And if the latter is the case, how would I differentiate the radios, say on a gateway between a cc110x network with `nomac` and an IEEE 802.15.4 network with `nomac`?

**Comment by authmillenon** [comment]
And how would I find out which address to use / which header (if we decide upon it) to use if the `NETMOD_NOMAC` option is true, or if the `NETMOD_CC110x` option is true, but the intermediate MAC layer has it's own header types.

**Comment by haukepetersen** [comment]
> It's also still unclear to me: If I have a e.g. a cc110x device with a nomac control thread, would I choose NETMOD_CC110x or NETMOD_NOMAC

You wouldn't use either of them :-) As interfaces are are accessed by there interface identifier, thats what you use. So if you have a standard IP address you want to send to, you ask your neighbor table to give you the interface id for that device. You ask then the netreg for the PID of that interface. In case of multicast/broadcast, the neighbor/forwarding table gives you a list of interface ID's you can lookup.

File: sys/net/include/netmod.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * @brief       List of all available network modules
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ */
+
+#ifndef __NETMOD_H_
+#define __NETMOD_H_
+
+
+typedef enum {
```
Comment (authmillenon) [inline sys/net/include/netmod.h]: Would start with something like `NETMOD_UNDEF` for unidentified data. To save space in the lookup-array and to have the correct behaviour for `NETMOD_NUMOF` I would suggest:

```C
typedef enum __attribute__((packed)) { // packed to make enum as small as possible
    NETMOD_UNDEF = -1,
#ifdef MODULE_UDP
    NETMOD_UDP,
#endif
/* … */
    NETMOD_NUMOF
}

File: sys/net/include/netmod.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#if MODULE_UDP
```
Comment (authmillenon) [inline sys/net/include/netmod.h]: `#ifdef`

File: sys/net/include/netmod.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    NETMOD_UDP,
+#endif
+#if MODULE_TCP
```
Comment (authmillenon) [inline sys/net/include/netmod.h]: `#ifdef`

File: sys/net/include/netmod.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    NETMOD_TCP,
+#endif
+
+#if MODULE_IPV6
```
Comment (authmillenon) [inline sys/net/include/netmod.h]: `#ifdef`

File: sys/net/include/netmod.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    NETMOD_IPV6,
+#endif
+#if MODULE_IPV4
```
Comment (authmillenon) [inline sys/net/include/netmod.h]: `#ifdef`

File: sys/net/include/netmod.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    NETMOD_IPV4
+#endif
+#if MODULE_6LOWPAN
```
Comment (authmillenon) [inline sys/net/include/netmod.h]: `#ifdef`

File: sys/net/include/netmod.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    NETMOD_6LOWPAN
+#endif
+
+#if MDULE_IEEE802154
```
Comment (authmillenon) [inline sys/net/include/netmod.h]: `#ifdef` + Typo

File: sys/net/include/netmod.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    NETMOD_IEEE802154
+#endif
+
+    /* ... you get the idea ... */
+
+    NETMOD_NUMOF        /**< shortcut for referencing the number of active protocols */
+} netmod_t;
```
Comment (authmillenon) [inline sys/net/include/netmod.h]: would suggest (in order of preference) `netproto_t` or `netproto_t` as type name. I associate `netmod_t` personally more with "network modification" than "network module".

File: sys/net/crosslayer/netreg/netreg.c [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    else {
+        newone = next_free;
+        newone->next = NULL;
+        next_free = next_free->next;
+    }
+
+    /* add to correct list */
+    do {
+        tmp = modules[module];
+    } while (tmp->next != NULL);
```
Comment (authmillenon) [inline sys/net/crosslayer/netreg/netreg.c]: This loop does not terminate

**Comment by haukepetersen** [comment]
as I said, it's more or less pseudo-code. I will only start to make it work when we really decide to run with this concept...

**Comment by OlegHahm** [comment]
Can we close this after #2404 got merged or do you plan to use this for a somehow optimized version of netreg?

**Comment by haukepetersen** [comment]
this PR is it's obsolete.

**Pull request closed without merging**